Do not completely remove ~/.rustup/tmp and download

ijackson opened the pull request:
As discussed in #2430, it is better to keep the top-level directory.  It might be a mountpoint, or a symlink which the user wants to keep, or have unusual permissions.

<strike>I wasn't able to find an implementation of "remove contents of this directory but not the directory itself", so I have implemeted one. Maybe this should be in some other crate somewhere but it's not that huge so I just put it here.</strike>  I use the new `remove_dir_contents` in the `remove_dir_all` crate

This change changes the behaviour of `delete_dir_contents` to actually just remove the contents, like the name suggests.  There are two call sites: one to delete tmp and one to delete download.  This is correct behaviour for both.

Fixes: #2430

I have done ad-hoc testing for this, and rerun the main test suite.  If you want a formal fails-before-passes-after test, I could provide one.

rbtcollins commented on src/utils/utils.rs:
```diff
@@ -627,12 +627,26 @@ where
+    // Deletes only the contents, and not the directory iteself (in
+    // case it's a mountpoint, or a symlink, or has unusual
+    // permissions which the user wants to preserve).
+    use anyhow::Context;
```
please put this use at the top of the file

rbtcollins commented on src/utils/utils.rs:
```diff
@@ -627,12 +627,26 @@ where
-    match remove_dir_all::remove_dir_all(dir_path) {
-        Err(e) if e.kind() != io::ErrorKind::NotFound => {
-            panic!("Unable to clean up {}: {:?}", dir_path.display(), e);
+    // Deletes only the contents, and not the directory iteself (in
```
I think this comment would make a lot of sense as a docstring - 
```
/// delete_dir_contents deletes ...
pub fn dele...
```

rbtcollins commented on src/utils/utils.rs:
```diff
@@ -627,12 +627,26 @@ where
+        .context("checking directory")?;
+        for entry in entries {
+            let entry_path = entry.context("reading directory")?.path();
+            remove_dir_all::remove_dir_all(&entry_path)
```
rather than reading the dir here and recursing into remove_dir_all, please patch remove_dir_all directly; inspecting https://github.com/XAMPPRocky/remove_dir_all/blob/master/src/fs.rs#L36 will show you why.  I have commit and publish rights on it, so can ensure a good response time.

I'm slightly  in favour of a new method e.g. remove_dir_contents_all in that crate rather than a parameter to the existing function (but probably refactor the contents to use a common internal helper).

ijackson commented:
Robert Collins writes ("Re: [rust-lang/rustup] Do not completely remove ~/.rustup/tmp and download (#2433)"):
> @rbtcollins requested changes on this pull request.

Hi.  Thanks for the review.  Your requested changes look good to me
and I'll go ahead and dothat.

I have a few opinions/questions about this comment:

> In src/utils/utils.rs:
> 
> >          }
> -        _ => {}
> -    }
> +        .context("checking directory")?;
> +        for entry in entries {
> +            let entry_path = entry.context("reading directory")?.path();
> +            remove_dir_all::remove_dir_all(&entry_path)
> 
> rather than reading the dir here and recursing into remove_dir_all, please
> patch remove_dir_all directly; inspecting https://github.com/XAMPPRocky/
> remove_dir_all/blob/master/src/fs.rs#L36 will show you why. I have commit and
> publish rights on it, so can ensure a good response time.

Thanks.  I'm very happy to make this change in the dependency.

Having read all that stuff in remove_dir_all, it seems to me that on
Windows we ought to remove each of the subdirectories separately.

That way when we move the contents to temporary names, the temporary
names are *inside* the toplevel directory (which we are not removing).
That will work better if the toplevel directory is a mount point or
junction or whatever these Windows things are called :-).

Do you agree ?

If so I will implement it that way.  I'll put this rationale in a
comment (or in the commit message if you prefer) for future reference.

> I'm slightly in favour of a new method e.g. remove_dir_contents_all in that
> crate rather than a parameter to the existing function (but probably refactor
> the contents to use a common internal helper).

I agree with your opinion that the public API should be a new
`remove_dir_contents_all` function.

I'll see what the innards come out looking like, but in any case I
will avoid duplication obviously, and do any refactoring needed to
support that.

Regards,
Ian.

-- 
Ian Jackson <ijackson@chiark.greenend.org.uk>   These opinions are my own.  

Pronouns: they/he.  If I emailed you from @fyvzl.net or @evade.org.uk,
that is a private address which bypasses my fierce spamfilter.

rbtcollins commented:
There's some additional prior art that may give possibly useful context on the fs behaviour https://github.com/cygwin/cygwin/blob/ad101bcb0f55f0eb1a9f60187f949c3decd855e4/winsup/cygwin/syscalls.cc#L675-L1064

>  Having read all that stuff in remove_dir_all, it seems to me that on Windows we ought to remove each of the subdirectories separately.

Do you mean something like 'base_dir should be the dir whose contents we are removing' ? Yes, that seems like an obvious choice since for this new function we don't have the same implied 'has permissions to delete the target' semantics.

ijackson commented:
Robert Collins writes ("Re: [rust-lang/rustup] Do not completely remove ~/.rustup/tmp and download (#2433)"):
> There's some additional prior art that may give possibly useful context on the
> fs behaviour https://github.com/cygwin/cygwin/blob/
> ad101bcb0f55f0eb1a9f60187f949c3decd855e4/winsup/cygwin/syscalls.cc#L675-L1064
> 
>     Having read all that stuff in remove_dir_all, it seems to me that on
>     Windows we ought to remove each of the subdirectories separately.
> 
> Do you mean something like 'base_dir should be the dir whose contents we are
> removing' ? Yes, that seems like an obvious choice since for this new function
> we don't have the same implied 'has permissions to delete the target'
> semantics.

Yes, precisely.

I'll look at the code again (quite tired now) but I suspect the most
convenient implementation of this going to look very similar to the
modified delete_dir_contents from my initial MR branch - it can call
remove_dir_all as a subroutine and then it will all DTRT.

Ian.

-- 
Ian Jackson <ijackson@chiark.greenend.org.uk>   These opinions are my own.  

Pronouns: they/he.  If I emailed you from @fyvzl.net or @evade.org.uk,
that is a private address which bypasses my fierce spamfilter.

mbrubeck commented:
@J-Vin, please don't spam PRs with comments or reviews that don't add new information.  It's very distracting. Thanks!

mbrubeck commented:
@J-Vin, you need to stop “reviewing” issues that you don’t own. It generates a lot of spam to our project members. We have blocked your account, for now.  Please contact rust-mods@rust-lang.org if you have any questions about this.

rbtcollins commented:
@ijackson this is now ready to be updated to remove_dir_all 0.7.0 and the new functions you wrote.

ijackson committed:
Update remove_dir_all to 0.7.0

Signed-off-by: Ian Jackson <ijackson@chiark.greenend.org.uk>

ijackson committed:
delete_dir_contents: Use remove_dir_contents

This is used for
 * In temp.rs, the "root directory" (ie .rustump/tmp)
 * In check_updates, on tmp and toolchains

We do not want to delete the containing directory, as discussed in
  https://github.com/rust-lang/rustup/issues/2430

There is still raw::remove_dir, which is used in a variety of places,
notably utils::remove_dir, some of the calls to which need fixing.

Signed-off-by: Ian Jackson <ijackson@chiark.greenend.org.uk>

ijackson commented:
@rbtcollins Thanks, now done :-).

kinnison reviewed (approved):
Looks good to me, if @rbtcollins is okay with this then we shall merge it after 1.24.2 release is done.
